File: 0017.LetterCombinationsofaPhoneNumber.go
func letterCombinations(digits string) []string {
    if len(digits) == 0 {
        return nil
    }
    key := []string{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
    // solution2
    ans := []string{""}
    for _, d := range digits {
        nxt := []string{}
        for _, path := range ans {
            for _, ch := range key[int(d-'0')] {
                nxt = append(nxt, path+string(ch))
            }
        }
        ans = nxt
    }
    /* solution1
    ans := []string{}
    dfs(digits, "", key, &ans)
    */
    return ans
}

func dfs(digits, path string, key []string, ans *[]string) {
    if len(digits) == 0 {
        *ans = append(*ans, path)
        return
    }
    for _, ch := range key[int(digits[0]-'0')] {
        dfs(digits[1:], path+string(ch), key, ans)
    }
}

File: 0000-0999/0017.LetterCombinationsofaPhoneNumber.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        dic = {'1':None, '2':'abc', '3':'def', '4':'ghi', '5':'jkl', '6':'mno', '7':'pqrs', '8':'tuv', '9':'wxyz', '0':' '}
        ans = []
        self.combin(dic, digits, 0, '', ans)
        return ans

    def combin(self, dic, digits, i, comb, ans):
        if i == len(digits):
            ans.append(comb)
            return
        for ch in dic[digits[i]]:
            self.combin(dic, digits, i+1, comb+ch, ans)